src/components/StandardForm.vue

```diff
@@ -0,0 +1,108 @@
+          v-if="col.selectItems && col.selectItems.length > 0"
+          class="form-control"
+          :id="col.field"
+          v-model="record[col.field]"
+          :readonly="col.readonly"
+          :class="classesForField(col.field)"
+          :aria-describedby="`feedback-${col.field}`"
+          >
+          <option v-for="opt in col.selectItems" :key="opt.value" :value="opt.value">
+            {{ opt.label }}
+          </option>
+        </select>
+
+        <input
+          v-else
+          type="text"
+          class="form-control"
+          :id="col.field"
+          v-model="record[col.field]"
+          :readonly="col.readonly"
+          :class="classesForField(col.field)"
+          :aria-describedby="`feedback-${col.field}`"
+          />
+
+        <div :id="`feedback-${col.field}`" class="invalid-feedback">
+          <span v-for="error in errors[col.field]" :key="error">
+            {{ error }}
+          </span>
+        </div>
+      </div>
+    </div>
+
+    <div class="form-group row">
+      <div class="col-sm-2"></div>
+      <div class="col-sm-10">
+        <button class="btn btn-primary" v-on:click="validateAndSave()" :disabled="disabled">Save</button>
+        <button class="btn btn-secondary ml-2" v-on:click="$emit('reset')" :disabled="disabled">Reset</button>
+      </div>
+    </div>
+  </div>
+</template>
+
+<script>
+export default {
+  name: 'StandardForm',
+  props: {
+    initialRecord: Object,
+    cols: Array,
+    /* example: {
+      label: 'Person name',
+      field: 'person_name',
+      readonly: false,
+      required: true,
+      selectItems: [
+        label: 'None', value: null,
+        label: 'Bob', value: 'Robert'
+      ]
+    } */
+    disabled: Boolean
+  },
+  data: function () {
+    return {
+      wasValidated: false,
+      record: Object.fromEntries(this.cols.map(col => [col.field, this.initialRecord[col.field]])),
+      errors: Object.fromEntries(this.cols.map(col => [col.field, []]))
+    }
+  },
+  methods: {
+    classesForField: function (field) {
+      const col = this.cols.find(col => col.field === field)
+      if (col.readonly) return {}
+      if (!this.wasValidated) return {}
+
+      return {
+        'is-valid': this.errors[field].length === 0,
+        'is-invalid': this.errors[field].length > 0
+      }
+    },
+    validateAndSave: function () {
+      this.cols.forEach(col => {
+        const field = col.field
+        const val = this.record[field]
+
+        // clear errors; we will re-set them if they still exist
+        this.errors[field] = []
+
+        // check required fields
+        if (col.required && (val === undefined || val === null || val === '')) {
+          this.errors[field].push(`${col.label} is a required field`)
+        }
+      })
+
+      this.wasValidated = true
+      const hasErrors = Object.values(this.errors).some(errList => errList.length > 0)
+      if (!hasErrors) {
+        this.$emit('save', this.record)
```

Comment:
Could you explain the advantage of events over callbacks is, as you see it?

It seems very much the same thing to me, only with an extra level of asynchrony introduced.

Reply:
i think it makes it easier, in a parent component, to see when functions are being called. it makes sense to me that the parent component is responsible for the flow of data, and the form is just responsible for layout, ui, and emitting events